### ISSUE TITLE
reproducibility of TurnKey packages

### ISSUE THREAD

**Issue by JedMeister** [turnkeylinux/tracker#1040]
This issue is to double-check reproducibility of all TurnKey packages and where need be discuss what might be required to ensure reproducibility.

----

All TurnKey packages. Ticked means currently reproducible. Tested by comparing locally built package against package downloaded from the TurnKey apt repo (recently built and uploaded by Alon).

**Diffoscope links** are provided for non-reproducible packages. Alon's package (from the [stretch repo](http://mirror.turnkeylinux.org/turnkeylinux/apt/debian/pool/stretch/main/)) on the left, with my locally build package on the right.

 - [ ] autoversion [1] - (see [diffoscope](https://try.diffoscope.org/byfvedupuzrs.html))
 - [x] busybox-initramfs
 - [x] casper
 - [x] ccurl
 - [ ] chanko [1] - (see [diffoscope](https://try.diffoscope.org/echbsebyqqcr.html))
 - [x] cloudtask
 - [x] confconsole
 - [ ] deck - currently missing from the repo - package needs work! (@jedmeister to do)
 - [ ] deckdebuild [2] - (see [diffoscope](https://try.diffoscope.org/pgdbudzzhsgh.html))
 - [x] di-live
 - [x] ebsmount
 - [x] ec2metadata
 - [ ] fab [1] - (see [diffoscope](https://try.diffoscope.org/cvefpxfvkfrp.html))
 - [x] hubclient
 - [x] hubdns
 - [x] hubtools
 - [x] inithooks
 - [ ] pool [1] - (see [diffoscope](https://try.diffoscope.org/tarehqffcxtg.html))
 - [x] pycurl-wrapper
 - [ ] pyproject-common [3] - Alon needs to pull tags, then should be ok (see [diffoscope](https://try.diffoscope.org/rebayccffsyr.html))
 - [x] python-dialog
 - [ ] repo [1] - (see [diffoscope](https://try.diffoscope.org/edkxcxuqvrhr.html))
 - [x] tklamq
 - [x] tklbam
 - [ ] tklbam-duplicity [4] - @jedmeister to double-check source for this one as my commit IDs differ to Alon's
 - [ ] tklbam-python-boto [4] - @jedmeister to double-check source for this one as my commit IDs differ to Alon's
 - [x] tklbam-squid
 - [x] tklbam-squid-common
 - [x] tklpatch
 - [x] turnkey-pylib
 - [x] turnkey-sysinfo
 - [x] turnkey-version
 - [ ] verseek [1] - (see [diffoscope](https://try.diffoscope.org/bcykypggqygc.html))
 - [x] webmin [and all standard modules] - note **webmin-webmincron** missing from repo.
 - [ ] webmin-tklbam [5] - (see [diffoscope](https://try.diffoscope.org/pxrzksufyufr.html))

----
Comments
-------------

[1] autoversion, chanko, fab, pool, repo and verseek 

 - all appear to be similar in as much as they all have a slightly different "installed size" (Alon's slightly smaller than mine)
 - also the control file notes a depend on a different version of libc6 (Alon: ​2.​4 ; mine: ​2.​2.​5).

However, I note that there are also other differences too. E.g.:
 - autoversion, chanko, pool, repo & verseek
   - besides different file sizes for `​/​usr/​bin/​<app_name>`, my packages also appear to contain an empty `​/​usr/​lib/​<app_name>/​libexec/​` dir
 - verseek
   - `/​usr/​share/​doc/​verseek/​RelNotes-​1.​0.​txt.​gz` & `/​usr/​share/​doc/​verseek/​copyright` have different timestamps
 - fab
   - appears to have the files sorted in a different order in `data.tar`, as well as slightly different timestamps

----
[2] deckdebuild

- empty `/​usr/​lib/​deckdebuild/​libexec/​` dir

----
[3] pyproject-common

- as noted, Alon needs to pull the tags so the version matches. Hopefully that will resolve that (diffoscope suggests so by my reading)

----
[4] tklbam-duplicity & tklbam-python-boto

- as noted, our packages have differing commit ids. Needs further investigation on my end.

----
[5] webmin-tklbam

- `/​usr/​share/​doc/​webmin-​tklbam/​copyright` has different timestamps?! Out by about 4 days by my reading (weird?!)

**Comment by lamby** [comment]
The first think I would do is check you are really building from the same source environment before drilling down into the weeds that are very likely just *symptoms* caused by things further up the chain.

For example, do not fixate at all on "installed size" being different. It's merely a summary of the symptom. This is like trying to "fix" the `sha1sum` binary to make something reproducible and you can waste a lot of time trying to force some small thing to be the same in a package (and not achieving actual reproduciblity) whilst zooming out at the bigger picture would have solved everything in one go.

So anyway I'm seeing stuff like ​`libc6·​(>=·​2.​4)` vs `libc6·​(>=·​2.​2.​5)​` which my gut tells me that someone is building in an almost totally-different environment (Different Debian distribution? Tainted chroot? Chroot with weird APT pinning? Three possible suggestions of many...)  I mean, if the build environment can't agree on which libc6 the final packages should use, I would guess that "all bets are off" for anything else to be reproducible and, indeed, if this implies that they are using different C compilers it is not surprising that you would be seeing differences in the binaries.

What might be useful at this point to compare the `.buildinfo` files between two builds. That would succinctly show what is different between the environments, including that the source is actually the same. It would also confirm the value of `SOURCE_DATE_EPOCH`. I would also check you are running the same command to build, especially around cleaning options (eg. `-nc`, etc.)

Regarding `usr/​lib/​deckdebuild/​libexec/​` existing in one build and not the other. Note that this is an empty directory, something that git cannot actually store (without hacks). Thus if someone is building from Git and someone isn't.. you see the problem I trust. Just an idea anyway.

(Regarding pyproject-common, you seem to be comparing two entirely different versions of the package there so I'm not sure what you were expecting?)

Regarding webmin-tklbam. It smells like that the copyright file is generated on build (or in deckdebuild?). This would suggest its mtime is not being set to `SOURCE_DATE_EPOCH` (or not the same value).

**Comment by JedMeister** [comment]
@lamby - brilliant; thanks mate! :smile: 

> The first think I would do is check you are really building from the same source environment before drilling down into the weeds that are very likely just symptoms caused by things further up the chain.

Alon has built his own buildroot. In theory, it should be the same as (or at least close to) mine, but may not be quite the same (I'll check with him). Also whilst it *shouldn't* matter (at least by my understanding), his host is Stretch, whereas mine (and @OnGle's) is Jessie, do you think that might have some impact?

> if the build environment can't agree on which libc6 the final packages should use, I would guess that "all bets are off" for anything else to be reproducible and, indeed, if this implies that they are using different C compilers it is not surprising that you would be seeing differences in the binaries.

Yes good call. Would different host cause this (as per above note)? If so, is that expected behaviour? I was under the impression that everything should be built within the chroot using whatever that has (otherwise how would you be able to build fro different releases?)

> What might be useful at this point to compare the .buildinfo files between two builds. 

Awesome, I'll ask Alon for his copies of those.

> Regarding `usr/​lib/​deckdebuild/​libexec/`​ existing in one build and not the other. Note that this is an empty directory, something that git cannot actually store (without hacks). Thus if someone is building from Git and someone isn't.. you see the problem I trust. Just an idea anyway.

Hmm. I'm definitely building from git (and I have the empty dirs) and AFAIK so is Alon, so I'm not 100% sure about that. Perhaps I'll do a clean clone from GH (as Alon may have done) and try again?

I'm assuming not, but is there any point/value in the empty dirs being there? (I.e. *should* they be there or not?)

> (Regarding pyproject-common, you seem to be comparing two entirely different versions of the package there so I'm not sure what you were expecting?)

Whilst on face value, they are completely different versions. But from what I can see (and I may be wrong), they're actually from the same commit, just Alon's doesn't have the git-tag. WIthout the git-tag, the version generated by autoversion is significantly different.

>Regarding webmin-tklbam. It smells like that the copyright file is generated on build (or in deckdebuild?). This would suggest its mtime is not being set to SOURCE_DATE_EPOCH (or not the same value).

Weird that I can reproduce on my local system and @OnGle can also reproduce (same as mine) on his, but Alon's is different...Thanks for the hin thought! It gives me a good place to start looking. 

Thanks again for the broader advice and context too. Muchly appreciated. I do want to have this resolved for the stable release, but as I say, I'll likely push on as-is for the Core RC.

**Comment by lamby** [comment]
> his host is Stretch, whereas mine (and @OnGle's) is Jessie

Indeed, it would very much cause differences. I mean, you are using different versions of (at least) the C compiler so it only seems natural that (for example) they would emit different machine code, possibly due to better optimisations or for other misc reasons. (That is the practical side of things, but from the "legal" angle, reproducible build is defined as given the same source code *and* the same build environment.)

It would also explain the `libc` dependency. These are automatically generated based on symbol versioning - if the binary uses some internal feature `X` that is only present in `libc6` version `Y` then it will naturally generate a `.deb` package that has a `Depends` of `libc6 (>= Y)`. Even the same source can generate different symbol versions if the compiler (which is different between `jessie` and `stretch` as mentioned above) feels the need to (for example, hardening flags, or updated math foo... but many other reasons too.)

> Would different host cause this

Oh hang on, to clarify, do you mean the machine on which the build is happening with a chroot *inside* that? If so, then the host machine should have nothing to do with it outside of some extremely esoteric cases that I've found involving `uname` calls that leaked data from the outside environment (unlikely to be the case here and probably not the first place to start).

If not, then I might repeat my remarks about ensuring the chroot/buildroot really is identical in any meaningful sense, and the easiest way of doing that is to compare .buildinfo files. Indeed, diffoscope will take two `.buildinfo` files as arguments.

> [pyproject-common] are different versions. But from what I can see (and I may be wrong), they're actually from the same commit, just Alon's doesn't have the git-tag. WIthout the git-tag, the version generated by autoversion is significantly different.

Then in that case the package will be once you ensure that you fixup the tag foo, as all the differences are related to the version number (see https://try.diffoscope.org/rebayccffsyr.html). Unless I'm missing something, of course.

**Comment by JedMeister** [comment]
> do you mean the machine on which the build is happening with a chroot inside that? 

Yes ^^ that ^^

FWIW the buildroot is built very much like any other TurnKey appliance: https://github.com/turnkeylinux/buildroot

> If so, then the host machine should have nothing to do with it outside of some extremely esoteric cases that I've found involving uname calls that leaked data from the outside environment (unlikely to be the case here and probably not the first place to start).

That's what I thought. To reiterate, both Alon and I are using a Stretch buildroot chroot to build packages. It's just the host system (that the buildroot chroot is running on) that differs.

Although I have just noticed, that Alon obviously didn't realise that we have a `stretch-dev` branch (of the buildroot) that we've been working from. He's done his own updates and pushed his changes to master. 

His changes are mostly inline with what we'd been using. The only change that looks like it may have some significance is he's added the `-d` switch to `dpkg-buildpackage`. The dpkg-buildpackage help notes: `-d    do not check build dependencies and conflicts.` Whilst, I'm not completely convinced that would the cause of difference (nor whether that's even what we'd really want), I'm not sure. Either way, it's probably good if we're definitely using the same buildroot. So I'll rebuild my buildroot and try again...

One other thing that may be relevant (which I'd previously overlooked) is that he almost certainly had newer versions of the packages turnkey-pylib, pyproject & autoversion. I'll make sure I update them too before retrying.

> Then in that case the package will be once you ensure that you fixup the tag foo, as all the differences are related to the version number (see https://try.diffoscope.org/rebayccffsyr.html). Unless I'm missing something, of course.

Agreed! FWIW, from the OP:

> pyproject-common [...] Alon needs to pull the tags so the version matches. Hopefully that will resolve that (diffoscope suggests so by my reading)

Thanks again @lamby :+1:

**Comment by lamby** [comment]
> His changes are mostly inline with what we'd been using.

You can probably bypass a lot of tedious checking here by simply comparing the `.buildinfo` files. That is the canonical description of what, exactly, was used as part of the build; no need to investigate hypothesis.

The `-d` switch should not do anything whatsoever, although potentially if one build was not installing an (optional) build-dependency, that could result in the package building something different as it might use a feature if it's available. Somewhat unlikely but again, just check the `.buildinfo` files; so much simpler.

> One other thing that may be relevant (which I'd previously overlooked) is that he almost certainly had newer versions of the packages turnkey-pylib, pyproject & autoversion. I'll make sure I update them too before retrying.

`.buildinfo` would have saved you having to think of this and/or rule it out completely. :)

**Comment by JedMeister** [comment]
>  checking here by simply comparing the .buildinfo files.

> again, just check the .buildinfo files; so much simpler.

> .buildinfo would have saved you having to think of this and/or rule it out completely. :)

Yep! I'll ask Alon for his build logs and take it from there... :smile:

**Comment by lamby** [comment]
Looking forward to it, but just to clarify build logs != `.buildinfo`..

**Comment by JedMeister** [comment]
So now that I have rebuilt my buildroot and refreshed the package source, the issues for most packages have just disappeared! The ones that don't match Alon's are definitely reproducible locally, they just don't match Alon's.

Re the `.buildinfo`, I spoke with Alon about that, it it appears that currently they are being destroyed as a matter of course. So we should be making sure they get kept!

It did occur to me that it might be a good idea to bundle a buildroot for users to build against?! Although I guess that sort of defeats the whole point in some regards, doesn't it?! Perhaps a better option is to publish the package versions of the buildroot? (perhaps the `.buildinfo` includes some/all of that info?)

If you have any suggestions on how we can make this work better, I'd love to hear

**Comment by lamby** [comment]
> So now that I have rebuilt my buildroot and refreshed the package
> source, the issues for most packages have just disappeared
[..]
> The ones that don't match Alon's are definitely reproducible locally,
> they just don't match Alon's.

Right, you've got to use the same toolchain and sources, after all!

> Perhaps a better option is to publish the package versions of the
> buildroot? (perhaps the .buildinfo includes some/all of that info?)

Hang on, have you seen a `.buildinfo` file? You can generate them
yourself quite easily but here is an example:

  https://buildinfo.debian.net/a6bc4c0f83bb78f0b2960c7fe610148b74e02eff/opari_1.1+dfsg-5_i386.buildinfo
  
Note how it includes the installed build dependencies, the environment, the sources, checksums... exactly all of the things needed to debug these issues, hence why I so keen to obtain them in previous replies...

I have a collection of over 6 million of these files here, in case you want any any more:

  https://buildinfo.debian.net/

**Comment by JedMeister** [comment]
We need to review this and ensure that all packages are reproducible. However, I'm moving it to v16.1 as I don't want it to block the upcoming and well overdue v16.0 release.

**Comment by OnGle** [comment]
Well we ran out of time, let's try again for 17.0
